examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,243 @@
+        self.searcher_metric = searcher_config["metric"]
+        self.searcher_unit = list(searcher_config["max_length"].keys())[0]
+        self.searcher_max_length = list(searcher_config["max_length"].values())[0]
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+        self._check_searcher_compatibility(args)
+
+    def on_log(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        logs=None,
+        **kwargs,
+    ):
+        if state.is_world_process_zero:
+            metrics, metric_type = self._get_metrics(logs)
+            if metric_type == TRAIN:
+                # Prevents reporting metrics for the same step twice. This happens after
+                # training is completed and average training metrics are reported with
+                # the same step as the in-progress training metrics.
+                if self.last_metrics["train_step"] != state.global_step:
+                    self.core_context.train.report_training_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["train_step"] = state.global_step
+
+            elif metric_type == EVAL:
+                # Prevents reporting metrics for the same step twice. This happens when
+                # after-training evaluation is completed, and it is reported with the same
+                # step as the last during-training evaluation.
+                if self.last_metrics["eval_step"] != state.global_step:
+                    self.core_context.train.report_validation_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["eval_step"] = state.global_step
+                    self.last_metrics.update(metrics)
+            else:
+                logging.warning(f"Metrics not reported: metric type = {metric_type}.")
+
+            self.last_metrics.update(metrics)
+
+        if self.core_context.preempt.should_preempt():
+            control.should_save = True
+
+    def _get_metrics(self, logs: typing.Dict) -> typing.Tuple[typing.Dict, str]:
+        metrics = logs
+        metric_type = get_metric_type(logs)
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics, metric_type
+
+    def on_save(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        info = det.get_cluster_info()
+        assert info
+
+        # local_path is where HF Trainer saves model and tokenizer.
+        local_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
```

Comment:
are there output files in the top level `output_dir` that are always generated?  would a user want to persist those in the checkpoint as well?  i.e. would we want to save any file in the top level dir along with the files in the checkpoint folder for a particular `global_step`?

Reply:
good point. `output_dir` contains two types of dirs:
* `checkpoint-step` with all files necessary for resumption
* `runs\date_ip\events.out.tfevents...`  which are events that can be visualized in the Tensorboard. It could be useful to some users and we should save it.

This means saving top-level `output_dir`. One issue is that `output_dir` contains all checkpoints which means its size will be gradually growing. Users can solve that problem for us by using `--save_total_limit N`, which limits number of checkpoints in `output_dir` to N. We cannot really solve that problem for users, since upload/store_path does not have a selector, which means we upload all content of the directory. A possible solution would be to use `store_path` and copy `runs` and selected `checkpoint_dir` manually, but I think I would rather rely on `--save_total_limit N` and copy full `output_dir` than to show users incorrect usage of `store_path` with manual copying.

Reply:
You can use a trick like https://github.com/determined-ai/determined/blob/42e91783665ba18746bc10bdcc02809e46912cca/examples/pytorch_lightning/core_api/integration.py#L178 to `upload` a selection of files without an additional copy by making symbolic links.

That said, maybe we should set `TensorboardMode` to `MANUAL` and upload the tensorboard artifacts that way?  That way they show up as expected in the Determined Tensorboard.  https://github.com/determined-ai/determined/blob/master/harness/determined/core/_tensorboard_mode.py

Reply:
That's a helpful trick I haven't thought about, thanks Corey! 
Still, I am looking into adding a selector to `upload` to avoid doing that, but if I cannot make it happen within a reasonable timeline, I am going to follow your suggestion.